www/docs/server/validators.md

```diff
@@ -0,0 +1,222 @@
+        code: 'FORBIDDEN',
+        message: "We don't take kindly to out-of-town folk",
+      });
+    }
+
+    return opts.next();
+  });
+
+export const appRouter = t.router({
+  hello: procedureWhichRejectsOutOfTownFolk
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const input = opts.input;
+      //    ^?
+      return {
+        greeting: `Hello ${input.name}, my friend`,
+      };
+    }),
+});
+```
+
+### Output Validators
+
+By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.
+
+Validators are used to infer the input and output types of your procedures.
+
+:::info
+If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
+:::
+
+## Validator integrations
+
+tRPC works out of the box with a number of popular validation and parsing libraries. The below are some examples of usage with validators which we officially maintain support for.
+
+### With [Zod](https://github.com/colinhacks/zod)
+
+Zod is our default recommendation, it has a strong ecosystem which makes it a great choice to use in multiple parts of your codebase. If you have no opinion of your own and want a powerful library which won't limit future needs, Zod is a great choice.
+
+```ts twoslash
+import { initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+
+export const appRouter = t.router({
+  hello: t.procedure
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .output(
+      z.object({
+        greeting: z.string(),
+      }),
+    )
+    .query(({ input }) => {
+      //      ^?
+      return {
+        greeting: `hello ${input.name}`,
+      };
+    }),
+});
+
+export type AppRouter = typeof appRouter;
+```
+
+### Writing a custom validator
+
+If you like, you can simply pass a function as a validator, though this isn't necessarily the best choice to use widely unless you have a specific need.
+
+```ts twoslash
+import { initTRPC } from '@trpc/server';
+
+export const t = initTRPC.create();
+
+export const appRouter = t.router({
+  hello: t.procedure
+    .input((value: any): { name: string } => {
+      if (value && typeof value.name === 'string') {
+        return { name: value.name };
+      }
+      throw new Error('Greeting not found');
+    })
+    .output((value: any) => {
+      if (value && typeof value.greeting === 'string') {
+        return { greeting: value.greeting };
+      }
+      throw new Error('Greeting not found');
+    })
+    .query(({ input }) => {
+      //      ^?
+      return {
+        greeting: `hello ${input.name}`,
+      };
+    }),
+});
+
+export type AppRouter = typeof appRouter;
+```
+
+### With [Yup](https://github.com/jquense/yup)
```

Comment:
These could be tabs like they are on Quickstart?

Reply:
Oh I had a look at this and felt that actually the way things are is quite nice. Having a lot of linkable sections in the sidebar should pay us back later in visibility and linkability.

Reply:
Plus the list if quite long and may not scale well to tabs as it continues to grow